### PR TITLE
Cs/sc 3360 support road network algorithm

### PR DESCRIPTION
**Pull request by Charl1996**
## Product Description
[Demo video](https://vimeo.com/905617454/0f7b334709?share=copy)

The disbursement algorithm, "Road Network Algorithm",  makes use of the Mapbox Matrix API. This PR adds the ability for a project to add their own mapbox token on the Geospatial Settings page. 

image

Once the FF is enabled for a project, the new disbursement algorithm will appear on the Geospatial Settings page. Once the algorithm is selected, two new HTML elements will appear on the screen:
1) Text (password) input prompting the user for their mapbox token.
2) A button ("Test API Key") which the user can click to make a sample request to Mapbox in order to verify that the token works.

Note that the user don't have to click the "Test API Key" button in order to save the token; the button exists purely to test the token. 

## Technical Summary
[Ticket](https://dimagi-dev.atlassian.net/browse/SC-3360)

Noteworthy items:
1) The api token is stored in postgres as an encrypted string, but returned to the front end in plaintext (although rendered as a password) when loading the Settings page.
2) The "Test API Key" button sample API call happens purely in the browser; I decided we don't have to have a dedicated endpoint just for the server to make an API call to Mapbox if we could simply use ajax.  

## Feature Flag
New FF: `SUPPORT_ROAD_NETWORK_DISBURSEMENT_ALGORITHM`

### Safety story
Tested this feature on my local machine.

### Automated test coverage
- Tests for view
- Tests for model

### QA Plan
QA planned

### Migrations
- [x] The migrations in this code can be safely applied first independently of the code

### Rollback instructions
- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Review by zandre-eng — COMMENTED** [review]
Looks good, just a few minor comments.

File: corehq/apps/geospatial/models.py [hunk 1]
```diff
@@ -58,9 +59,10 @@ class GeoConfig(models.Model):
 
     selected_disbursement_algorithm = models.CharField(
         choices=VALID_DISBURSEMENT_ALGORITHMS,
-        default=ROAD_NETWORK_ALGORITHM,
+        default=RADIAL_ALGORITHM,
         max_length=50
     )
+    _api_token = models.CharField(max_length=255, blank=True, null=True, db_column="api_token")
```
Comment (SmittieC) [inline corehq/apps/geospatial/models.py:65]: Thoughts on using the [django_cryptography](https://django-cryptography.readthedocs.io/en/latest/examples.html) library? Seems like it could save us some code

File: corehq/apps/geospatial/static/geospatial/js/geo_config.js [hunk 2]
```diff
@@ -46,6 +50,33 @@ hqDefine("geospatial/js/geo_config", [
...
         });
 
+        self.selectedAlgorithm = ko.observable();
+        self.apiToken = ko.observable(data.api_token);
+
+        self.captureApiToken = ko.computed(function () {
+            return self.selectedAlgorithm() === ROAD_NETWORK_ALGORITHM;
+        });
+
+        self.validateApiToken = function () {
```
Comment (SmittieC) [inline corehq/apps/geospatial/static/geospatial/js/geo_config.js:60]: I know for connection settings the user can choose to test the connection before saving it, but personally I don't think we should follow that approach here. Reason being:
- The validity of the connection settings can change anytime from `valid` to `invalid` to `valid` again.
- The token's validity cannot change from invalid to valid, but can change valid to invalid

I suggest we test the token's validity when the user clicks save. If the token is valid, we continue, otherwise we tell the user it's not valid.

Comment (mkangia) [reply corehq/apps/geospatial/static/geospatial/js/geo_config.js:60]: > I suggest we test the token's validity when the user clicks save. If the token is valid, we continue, otherwise we tell the user it's not valid.

That does sound good.

Just one place we might block the user is, if they wish to change the token but the mapbox api is down and so the validation is unable to run.
I assume we will only run the test if the value has changed and not otherwise. That logic will need to be added as well.

So, there are pros and cons for both. Though I'd like if users would test the token if they change it.

File: corehq/apps/geospatial/static/geospatial/js/geo_config.js [hunk 3]
```diff
@@ -46,6 +50,33 @@ hqDefine("geospatial/js/geo_config", [
...
+            const url = "https://api.mapbox.com/directions-matrix/v1/mapbox/driving/-122.42,37.78;-122.45,37.91;-122.48,37.73";
+            const params = "access_token=" + self.apiToken();
+
+            $.ajax({
+                method: 'get',
+                url: url + "?" + params,
```
[the remark this answers is not in the document]
Comment (mkangia) [reply corehq/apps/geospatial/static/geospatial/js/geo_config.js:66]: Good point. How is the token added for other mapbox requests?

File: corehq/apps/geospatial/tests/test_views.py [hunk 4]
```diff
@@ -162,6 +162,61 @@ def test_config_update(self):
         self.assertEqual(config.target_group_count, 10)
         self.assertEqual(config.selected_disbursement_algorithm, GeoConfig.RADIAL_ALGORITHM)
 
+    @flag_enabled('GEOSPATIAL')
+    def test_config_update__road_network_algorithm_ff_disabled(self):
```
Comment (mkangia) [inline corehq/apps/geospatial/tests/test_views.py:166]: interesting use of double underscore for nesting these.

**Review by mkangia — CHANGES_REQUESTED** [review]
Quick question @Charl1996,
why are we putting this behind a new feature flag? Can't we simply include it in the feature itself.
If its new, can we simply add "(unstable)" or something similar in the name in the dropdown when user is selecting it to let them know it is still in development.
Since we already have a feature flag for GEOSPATIAL, this is anyway being used by known projects.

File: corehq/apps/geospatial/static/geospatial/js/geo_config.js (hunk 2 above)
Comment (SmittieC) [reply corehq/apps/geospatial/static/geospatial/js/geo_config.js:60]: I figure we'd know when Mapbox is down from the status we get back. In that case we 1. tell the user and 2. proceed and fall back on the logic that tells the user about the invalid token when they run the algorithm.

Comment (mkangia) [reply corehq/apps/geospatial/static/geospatial/js/geo_config.js:60]: Right.
I was just comparing the benefits with the amount of code this code add. Feel free to give it a go and see how you feel about it. I do see an advantage of verifying the token before its saved, manually by the user or automatically by HQ itself.

File: corehq/apps/geospatial/forms.py [hunk 5]
```diff
@@ -73,13 +81,27 @@ class Meta:
...
+        widget=forms.PasswordInput(),
+    )
 
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
 
+        if toggles.SUPPORT_ROAD_NETWORK_DISBURSEMENT_ALGORITHM.enabled(self.domain):
+            choices = self.fields['selected_disbursement_algorithm'].choices
+            choices.append(self.ROAD_NETWORK_ALGORITHM_OPTION)
+            self.fields['selected_disbursement_algorithm'].choices = choices
```
[the remark this answers is not in the document]
Comment (Charl1996) [reply corehq/apps/geospatial/forms.py:103]: I initially did that but it didn't work. I have to set it explicitly.

If you look at the code around how `choices` are handled, you see the following:
> choices = property(_get_choices, _set_choices)

There's a getter and a setter. The getter returns a deep copy of the choices, so simply appending it won't work, since you're appending to a copy of the actual values and not the values themselves, hence you need to use the setter method.

[the remark this answers is not in the document]
Comment (zandre-eng) [reply corehq/apps/geospatial/forms.py:103]: Ah, sneaky! Thanks for the explanation, all good to keep as is then :+1:

File: corehq/apps/geospatial/models.py [hunk 6]
```diff
@@ -81,3 +83,18 @@ def disbursement_solver(self):
...
+
+    @property
+    def api_token(self):
+        if self._api_token and self._api_token.startswith(f'${ALGO_AES}$'):
+            ciphertext = self._api_token.split('$', 2)[2]
+            return b64_aes_decrypt(ciphertext)
+        return self._api_token
+
+    @api_token.setter
+    def api_token(self, value):
```
[the remark this answers is not in the document]
Comment (Charl1996) [reply corehq/apps/geospatial/models.py:95]: [Here's](https://www.stavros.io/posts/how-replace-django-model-field-property/) also an article explaining what's happening here.

Baiscally I want to intercept a particular field's getter/setter methods, hence using this approach. And like Manish said, it's how we do it in other parts of the codebase as well.

File: corehq/apps/geospatial/static/geospatial/js/geo_config.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Charl1996) [reply corehq/apps/geospatial/static/geospatial/js/geo_config.js:66]: The token always gets passed as part of the query params.

File: corehq/apps/geospatial/models.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (zandre-eng) [reply corehq/apps/geospatial/models.py:95]: Thanks for the explanation, good to leave as is then. I just wanted to make sure there wouldn't be any confusion with sharing the same name.

File: corehq/apps/geospatial/static/geospatial/js/geo_config.js [hunk 7]
```diff
@@ -46,6 +50,33 @@ hqDefine("geospatial/js/geo_config", [
...
+                success: function (data, status, request) {
+                    if (request.status === 200) {
```
[the remark this answers is not in the document]
Comment (Charl1996) [reply corehq/apps/geospatial/static/geospatial/js/geo_config.js:68]: According to the [docs](https://docs.mapbox.com/api/navigation/matrix/#matrix-api-errors) one will always receive either a 200 or >= 400 status codes, in which case the `fail` subscription block will trigger.

Maybe I should collapse the success block if? I do concur that if there's something that happens that is not a 200 status code but triggered the `success` flow it could be weird...but then again, theoretically it won't happen.

File: corehq/apps/geospatial/routing_solvers/pulp.py [hunk 8]
```diff
@@ -87,7 +86,10 @@ class RoadNetworkSolver(RadialDistanceSolver):
     Solves user-case location assignment based on driving distance
     """
 
-    def calculate_distance_matrix(self):
+    def calculate_distance_matrix(self, domain):
+        from corehq.apps.geospatial.models import GeoConfig
```
[the remark this answers is not in the document]
Comment (Charl1996) [reply corehq/apps/geospatial/routing_solvers/pulp.py:90]: Correct.

[the remark this answers is not in the document]
Comment (Charl1996) [reply corehq/apps/geospatial/routing_solvers/pulp.py:90]: Now that I think about it, maybe I should pass in the domain's config and not the domain. That would make sense.

File: corehq/apps/geospatial/static/geospatial/js/geo_config.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (zandre-eng) [reply corehq/apps/geospatial/static/geospatial/js/geo_config.js:68]: Since we know what to expect from Mapbox, I think it should be fine to completely remove this conditional block. I agree that it would be quite strange, and very unlikely, to receive something other than 200 for success.

File: corehq/apps/geospatial/models.py (hunk 1 above)
Comment (Charl1996) [reply corehq/apps/geospatial/models.py:65]: Just to avoid a conflict with the `api_token` property. I suppose I could've named it something else, but seeing as the token from the database (i.e. `_api_token`'s value) should never be used outside of the model class it made sense to name it as such.

File: corehq/apps/geospatial/static/geospatial/js/geo_config.js (hunk 2 above)
Comment (Charl1996) [reply corehq/apps/geospatial/static/geospatial/js/geo_config.js:60]: I initially thought of making the request on save to check validity, but then again I don't think we should "force" a sample check. We don't do that for connection settings so I think it's better that we don't do it for this also. 

The button technically don't have to exist and is purely for the user to make sure they copied the right value. If they are sure and don't see the need to test it that's up to them. 

If we make a sample request on save (even if it's only if the token value changed) that means there's technically a side-effect eating through their free budget on Mapbox; what if the users don't want that?

**Comment by Charl1996** [comment]
> why are we putting this behind a new feature flag? 

The reason is simply because there's additional costs and considerations involved when using this algorithm. I suppose we don't have to. Let met check with Mary.

File: corehq/apps/geospatial/static/geospatial/js/geo_config.js (hunk 2 above)
Comment (mkangia) [reply corehq/apps/geospatial/static/geospatial/js/geo_config.js:60]: >  their free budget on Mapbox; what if the users don't want that?

I doubt one request is going to hurt the number of free requests they have or even get noticed. So, we would have been fine, if we were to do it.

File: corehq/apps/geospatial/models.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mkangia) [reply corehq/apps/geospatial/models.py:95]: thanks for linking that @Charl1996 
It does seem we do a similar thing in the example I linked, with a slight difference

https://github.com/dimagi/commcare-hq/blob/b0f5643b36cf1ad4e4e66dfe458872d7278b9896/corehq/motech/models.py#L121-L132

Just to highlight the difference, is that the db column and the simultaneous property are encrypted and a separate property is added to handle the setter.
I can see only one advantage to that is that if anyone accesses `api_token`, which is something one would type when looking for `api_token` of the model object, one would always get the encrypted version. So for example, say in logs, it would be the encrypted token or while accidentally printing in console, or also in the view when rendering, avoiding accidental decryption and display of the token.
And thus forcing or being explicit that its the plaintext version of token being viewed at.

Even Django uses the same approach for its User model
```
class AbstractBaseUser(models.Model):
    password = models.CharField(_('password'), max_length=128)
```

So, if this was not a sensitive column, the implementation you shared would have been the right way to go about it but since this is a token we are probably better with keeping things explicit.

Let me know your thoughts on this.

Personally, I would explicitly decrypt only when necessary so the decrypted version is alive for the least time during code execution.

File: corehq/apps/geospatial/static/geospatial/js/geo_config.js (hunk 2 above)
Comment (Charl1996) [reply corehq/apps/geospatial/static/geospatial/js/geo_config.js:60]: > I doubt one request is going to hurt the number of free requests they have or even get noticed

I do agree, but for me it's a principle thing; I don't think there should be background side-effects on a seemingly unrelated page like "Settings". The user expects the token to be used only when it's actually necessary (running disbursement), unless explicitly actioned elsewhere.

**Review by zandre-eng — APPROVED** [review]
Thanks for all the effort on this @Charl1996, looks good!

File: corehq/apps/geospatial/models.py [hunk 9]
```diff
@@ -85,16 +85,16 @@ def disbursement_solver(self):
...
-    def api_token(self, value):
+    @plaintext_api_token.setter
+    def plaintext_api_token(self, value):
         if value and not value.startswith(f'${ALGO_AES}$'):
             ciphertext = b64_aes_encrypt(value)
-            self._api_token = f'${ALGO_AES}${ciphertext}'
+            self.api_token = f'${ALGO_AES}${ciphertext}'
         else:
-            self._api_token = None
+            self.api_token = None
```
Comment (mkangia) [inline corehq/apps/geospatial/models.py:100]: nit: I am thinking that this else block is implicitly resetting the api_token.

I would think that this function would encrypt and store the value and if it is not able to should it
1. do nothing
2. raise an error

I would prefer 2. More specifically:
```
if value is None:
    self.api_token = None
elif value and not value.startswith(f'${ALGO_AES}$'):
    # encrypt and store
else:
    raise Exception("Unexpected value received for plaintext api token")
```

Let me know what you think.

**Review by mkangia — COMMENTED** [review]
Thanks for the follow up on using the plaintext property @Charl1996 
Appreciate the flexibility.

I assume this has not been deployed on any environment so can we simply club the migrations to keep only one final migration.

File: corehq/apps/geospatial/forms.py [hunk 10]
```diff
@@ -97,11 +97,6 @@ class Meta:
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
 
-        if toggles.SUPPORT_ROAD_NETWORK_DISBURSEMENT_ALGORITHM.enabled(self.domain):
-            choices = self.fields['selected_disbursement_algorithm'].choices
-            choices.append(self.ROAD_NETWORK_ALGORITHM_OPTION)
-            self.fields['selected_disbursement_algorithm'].choices = choices
-
```
Comment (mkangia) [inline corehq/apps/geospatial/forms.py]: Nice!

File: corehq/apps/geospatial/models.py (hunk 9 above)
Comment (Charl1996) [reply corehq/apps/geospatial/models.py]: https://github.com/dimagi/commcare-hq/pull/34024/commits/335f4c805ee18af13a41257b632b77af93d05b58

The setter now also handles the case where you set a non-string value, say, `1234`. You also can't set an empty api_token. The token can only ever be either
1. A string value (not empty)
2. null

File: corehq/apps/geospatial/models.py [hunk 11]
```diff
@@ -93,8 +93,13 @@ def plaintext_api_token(self):
...
-            self.api_token = None
+        try:
+            if value is None:
+                self.api_token = None
+            elif value and not value.startswith(f'${ALGO_AES}$'):
+                ciphertext = b64_aes_encrypt(value)
+                self.api_token = f'${ALGO_AES}${ciphertext}'
+            else:
+                raise Exception("Unexpected value set for plaintext api token")
+        except AttributeError:
```
Comment (mkangia) [inline corehq/apps/geospatial/models.py:104]: Thanks for the improvement @Charl1996 
Can you share what will raise the exception `AttributeError`?

Additionally, possible to keep the two messages different? If one gets an error, it wont be clear where it came from.

Comment (mkangia) [reply corehq/apps/geospatial/models.py:104]: When raising exceptions its nice to be explicit what code would raise the exception. It's not clear to me here what will raise the exception `AttributeError`, though if only the necessary code is kept in the try block, its easier to locate.
